src/VisualStudio/Core/Def/Implementation/TaskList/ExternalErrorDiagnosticUpdateSource.cs

```diff
@@ -348,107 +440,120 @@ internal enum BuildProgress
 
         private sealed class InProgressState
         {
-            private int _incrementDoNotAccessDirectly = 0;
-
             private readonly ExternalErrorDiagnosticUpdateSource _owner;
-            private readonly HashSet<ProjectId> _projectsBuilt = new HashSet<ProjectId>();
-            private readonly HashSet<ProjectId> _projectsWithErrors = new HashSet<ProjectId>();
 
-            private readonly Dictionary<ProjectId, ImmutableHashSet<string>> _allDiagnosticIdMap = new Dictionary<ProjectId, ImmutableHashSet<string>>();
-            private readonly Dictionary<ProjectId, ImmutableHashSet<string>> _liveDiagnosticIdMap = new Dictionary<ProjectId, ImmutableHashSet<string>>();
+            private readonly ConcurrentDictionary<ProjectId, ImmutableHashSet<string>> _allDiagnosticIdMap = new ConcurrentDictionary<ProjectId, ImmutableHashSet<string>>();
+            private readonly ConcurrentDictionary<ProjectId, ImmutableHashSet<string>> _liveDiagnosticIdMap = new ConcurrentDictionary<ProjectId, ImmutableHashSet<string>>();
```

Comment:
ConcurrentDicts scare me a lot.  Generally because they're not really teh right data structure for the task (i.e. over a Dictionary+lock).  Is this use just to allow safe reading/writing of the dic from two entities?  or do you really intend to have multiple concurrent writers that cannot mutex on each other?

Reply:
I can switch back the original Dictionary + lock implementation if that feels better - given there are bunch of data structures in this type which are guaranteed serialized access, while these two were the only ones that seem to be possible to get accessed concurrently, I felt this change made the threading guarantees/documentation more explicit.